### PR TITLE
Script based hyperkube to avoid dependencies

### PR DESCRIPTION
**Pull request by dims**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
cmd/hyperkube is a magnet that drags in all kinds of dependencies. After lot of back and forth discussion in the code-organization subproject and attempts at proof-of-concept of a external repository for it we ended up with this alternative. The main reason being that a cmd/hyperkube will by definition need to vendor `k8s.io/kubernetes` which we do not want to support as a community (see #84558) 

Essentially in this PR:
- we add back a hyperkube container image
- we have a shell script that replaces the go based binary
- the new container images bundles all the binaries supported (yes, this increases the size of the container image)
- Just like before the script looks at its command line parameters and launches the correct binary.

Note that this variation of hyperkube image is for convenience to help folks migrate to specific binary / images and not depend on the kubernetes project supporting a single convenience binary (and image) for all the components. Please see #81760 for additional discussion and notes/recording of the recent code-organization meetings as well for more context.

**Which issue(s) this PR fixes**:

Related to #81760

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Simple script based hyperkube image that bundles all the necessary binaries. This is a equivalent replacement for the image based on the go based hyperkube command + image.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: cluster/images/hyperkube/Dockerfile [hunk 1]
```diff
@@ -14,5 +14,8 @@
 
 FROM BASEIMAGE
 
-# Copy the hyperkube binary
+# Cleanup all the soft links
+ADD binaries.tgz /usr/local/bin
```
Comment (liggitt) [inline cluster/images/hyperkube/Dockerfile]: how does this line clean up soft links?

Comment (dims) [reply cluster/images/hyperkube/Dockerfile]: @liggitt by replacing the files :)

Comment (dims) [reply cluster/images/hyperkube/Dockerfile]: once this gets in, i will follow up to remove stuff from the base image (and this should continue to work even after that)

Comment (liggitt) [reply cluster/images/hyperkube/Dockerfile]: > If `<src>` is a local tar archive in a recognized compression format (identity, gzip, bzip2 or xz) then it is unpacked as a directory.

ah

File: cluster/images/hyperkube/Makefile [hunk 2]
```diff
@@ -33,7 +37,9 @@ ifndef VERSION
     $(error VERSION is undefined)
 endif
 	cp -r ./* ${TEMP_DIR}
-	cp ${HYPERKUBE_BIN} ${TEMP_DIR}
+
+	tar -cvzf ${TEMP_DIR}/binaries.tgz  -C ${OUTPUT_PATH} kube-apiserver kube-controller-manager \
+		kube-proxy kube-scheduler kubectl kubelet
 
 	chmod a+rx ${TEMP_DIR}/hyperkube
```
Comment (liggitt) [inline cluster/images/hyperkube/Makefile]: is this still needed? do we need to do the same for the binaries we extracted?

File: cluster/images/hyperkube/hyperkube [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  kube-controller-manager
+  kube-proxy
+  kube-scheduler
+  kubectl                  kubectl controls the Kubernetes cluster manager
+  kubelet
+EOF
+  exit 0
+}
+
+function main() {
```
Comment (liggitt) [inline cluster/images/hyperkube/hyperkube]: cc @BenTheElder or @cblecker for bash sanity

**Comment by liggitt** [comment]
this looks roughly like what I expected. would like an ack on the release script bits from @kubernetes/sig-release and the bash bits from @BenTheElder or @cblecker (or alternates)

File: cluster/images/hyperkube/hyperkube [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    echo "$1: command not supported"
+    print_usage
+  fi
+  command=${1}
+  shift
+  if [[ -z "$(which ${command})" ]]; then
+    echo "${command}: command not found"
+    return 1
+  fi
+  exec ${command} "${@}"
```
Comment (liggitt) [inline cluster/images/hyperkube/hyperkube]: does this propagate error code out, and stdin in? what does this do if the named binary cannot be found in the path?

File: cluster/images/hyperkube/hyperkube [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+if [[ "${BASH_SOURCE[0]}" == "${0}" ]]; then
```
Comment (liggitt) [inline cluster/images/hyperkube/hyperkube]: should this be conditional? what should this script do if this is false?

File: cluster/images/hyperkube/Makefile (hunk 2 above)
Comment (dims) [reply cluster/images/hyperkube/Makefile]: the binaries are being zipped up here, the zip has the correct permission bits. so we should be fine. the +x on the script is insurance i think (from before)

File: cluster/images/hyperkube/hyperkube (hunk 3 above)
Comment (dims) [reply cluster/images/hyperkube/hyperkube]: Ack will ping them

File: cluster/images/hyperkube/hyperkube (hunk 4 above)
Comment (dims) [reply cluster/images/hyperkube/hyperkube]: i do have the `which` just before this `exec`, that should cover the issue where the binary may not be found in path.

yes, the `exec` ensures that the stdin/out/err and exit code line up properly

File: cluster/images/hyperkube/hyperkube (hunk 5 above)
Comment (dims) [reply cluster/images/hyperkube/hyperkube]: i can yank these lines out. they are typically used when the same file is used to source methods etc.

**Comment by dims** [comment]
/assign @BenTheElder @cblecker 

Ben, Christoph, Could you please look over the bash bits here? (redirect request from Jordan)

File: cluster/images/hyperkube/Dockerfile (hunk 1 above)
Comment (BenTheElder) [reply cluster/images/hyperkube/Dockerfile]: `ADD` is `COPY`'s magical friend :D

File: cluster/images/hyperkube/Makefile (hunk 2 above)
Comment (BenTheElder) [reply cluster/images/hyperkube/Makefile]: yeah, to avoid umask issues when building

File: cluster/images/hyperkube/hyperkube [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  command=${1}
+  shift
+  if [[ -z "$(which ${command})" ]]; then
+    echo "${command}: command not found"
+    return 1
+  fi
+  exec ${command} "${@}"
+}
+
+main "${@}"
```
Comment (BenTheElder) [inline cluster/images/hyperkube/hyperkube]: newline is definitely going to get format bikeshedded later

File: cluster/images/hyperkube/hyperkube [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  fi
+  if ! array_contains "$1" "${BINS[@]}"; then
+    echo "$1: command not supported"
+    print_usage
+  fi
+  command=${1}
+  shift
+  if [[ -z "$(which ${command})" ]]; then
+    echo "${command}: command not found"
+    return 1
```
Comment (BenTheElder) [inline cluster/images/hyperkube/hyperkube]: why not exit? we're using exit in print_usage

File: cluster/images/hyperkube/hyperkube (hunk 4 above)
Comment (BenTheElder) [reply cluster/images/hyperkube/hyperkube]: `exec` exists pretty much for this use case.
```
$ help exec
exec: exec [-cl] [-a name] [command [arguments ...]] [redirection ...]
    Replace the shell with the given command.
    
    Execute COMMAND, replacing this shell with the specified program.
    ARGUMENTS become the arguments to COMMAND.  If COMMAND is not specified,
    any redirections take effect in the current shell.
    
    Options:
      -a name	pass NAME as the zeroth argument to COMMAND
      -c	execute COMMAND with an empty environment
      -l	place a dash in the zeroth argument to COMMAND
    
    If the command cannot be executed, a non-interactive shell exits, unless
    the shell option `execfail' is set.
    
    Exit Status:
    Returns success unless COMMAND is not found or a redirection error occurs.
```

File: cluster/images/hyperkube/BUILD [hunk 8]
```diff
@@ -1,16 +1,44 @@
...
+
 multi_arch_container(
-    name = "hyperkube",
+    name = "image",
     architectures = SERVER_PLATFORMS["linux"],
     base = "@debian-hyperkube-base-{ARCH}//image",
+    cmd = [
+        "/bin/bash",
+        "-c",
+        "/hyperkube",
```
Comment (BenTheElder) [inline cluster/images/hyperkube/BUILD]: why not just `/hyperkube`?

File: cluster/images/hyperkube/Dockerfile [hunk 9]
```diff
@@ -14,5 +14,8 @@
...
+
+# Copy the shell script
 COPY hyperkube /hyperkube
```
Comment (BenTheElder) [inline cluster/images/hyperkube/Dockerfile]: should we set the entrypoint ?? we are in the bazel image

File: cluster/images/hyperkube/hyperkube (hunk 7 above)
Comment (dims) [reply cluster/images/hyperkube/hyperkube]: Done

File: cluster/images/hyperkube/BUILD (hunk 8 above)
Comment (dims) [reply cluster/images/hyperkube/BUILD]: Done.

File: cluster/images/hyperkube/Dockerfile (hunk 9 above)
Comment (dims) [reply cluster/images/hyperkube/Dockerfile]: Done

File: cluster/images/hyperkube/hyperkube (hunk 6 above)
Comment (dims) [reply cluster/images/hyperkube/hyperkube]: fixed.

**Comment by dims** [comment]
@liggitt @BenTheElder fixed up the bash stuff. this is ready now ( hopefully! )

File: build/lib/release.sh [hunk 10]
```diff
@@ -28,6 +28,7 @@ readonly RELEASE_STAGE="${LOCAL_OUTPUT_ROOT}/release-stage"
 readonly RELEASE_TARS="${LOCAL_OUTPUT_ROOT}/release-tars"
 readonly RELEASE_IMAGES="${LOCAL_OUTPUT_ROOT}/release-images"
 
+KUBE_BUILD_HYPERKUBE=${KUBE_BUILD_HYPERKUBE:-y}
```
Comment (BenTheElder) [inline build/lib/release.sh]: minor thought: should this still be y by default?

Comment (dims) [reply build/lib/release.sh]: for 1.17, yes.

**Comment by dims** [comment]
/hold cancel

thanks @BenTheElder

**Comment by dims** [comment]
@liggitt this is now ready to merge.

File: cluster/images/hyperkube/BUILD [hunk 11]
```diff
@@ -1,16 +1,42 @@
...
+        "//cmd/kube-proxy",
+        "//cmd/kube-scheduler",
+        "//cmd/kubectl",
+        "//cmd/kubelet",
+    ],
+)
+
 multi_arch_container(
-    name = "hyperkube",
+    name = "image",
```
Comment (liggitt) [inline cluster/images/hyperkube/BUILD]: was there a reason this changed names? anyone using the existing build target would have to change their scripts, right?

File: cluster/images/hyperkube/Dockerfile [hunk 12]
```diff
@@ -14,5 +14,10 @@
 
 FROM BASEIMAGE
 
-# Copy the hyperkube binary
+# Cleanup all the soft links
+ADD binaries.tgz /usr/local/bin
```
Comment (liggitt) [inline cluster/images/hyperkube/Dockerfile]: is /usr/local/bin in the $PATH for this image?

Comment (dims) [reply cluster/images/hyperkube/Dockerfile]: yes it is.

File: cluster/images/hyperkube/BUILD (hunk 11 above)
Comment (dims) [reply cluster/images/hyperkube/BUILD]: the name was conflicting with the name of the script. if you see the conformance image, we use `"image"`, so changed this to match. 

note that this is not used by release automation and i could not find any references anywhere else, so we should be fine.

Comment (liggitt) [reply cluster/images/hyperkube/BUILD]: ok

Comment (dims) [reply cluster/images/hyperkube/BUILD]: we use "image" in kubemark for example https://github.com/kubernetes/kubernetes/blob/master/cluster/images/kubemark/BUILD#L6

**Comment by justaugustus** [comment]
Thanks for working on this, @dims! :)

**Pull request merged**